pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,482 @@
+}
+
+func (m *forwardingRuleModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+func (a *forwardingRuleAdapter) Find(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting ComputeForwardingRule", "name", a.resourceID)
+
+	var err error
+	forwardingRule := &computepb.ForwardingRule{}
+	if a.Location == "global" {
+		req := &computepb.GetGlobalForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		forwardingRule, err = a.globalForwardingRulesClient.Get(ctx, req)
+		if err != nil {
+			if direct.IsNotFound(err) {
+				return false, nil
+			}
+			return false, fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.fullyQualifiedName(), err)
+		}
+	} else {
+		req := &computepb.GetForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
+			Project:        a.projectID,
+		}
+		forwardingRule, err = a.forwardingRulesClient.Get(ctx, req)
+		if err != nil {
+			if direct.IsNotFound(err) {
+				return false, nil
+			}
+			return false, fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.fullyQualifiedName(), err)
+		}
+	}
+	a.actual = forwardingRule
+	return true, nil
+}
+
+func (a *forwardingRuleAdapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	if a.projectID == "" {
+		return fmt.Errorf("project is empty")
+	}
+	if a.resourceID == "" {
+		return fmt.Errorf("resourceID is empty")
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating ComputeForwardingRule", "name", a.resourceID)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+
+	forwardingRule := ComputeForwardingRuleSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	forwardingRule.Name = direct.LazyPtr(a.resourceID)
+	forwardingRule.Labels = desired.Labels
+	// TODO(yuhou): handle default spec values
+	if forwardingRule.LoadBalancingScheme == nil {
+		forwardingRule.LoadBalancingScheme = direct.LazyPtr("EXTERNAL")
+	}
+
+	var err error
+	op := &gcp.Operation{}
+	if a.Location == "global" {
+		req := &computepb.InsertGlobalForwardingRuleRequest{
+			ForwardingRuleResource: forwardingRule,
+			Project:                a.projectID,
+		}
+		op, err = a.globalForwardingRulesClient.Insert(ctx, req)
+	} else {
+		req := &computepb.InsertForwardingRuleRequest{
+			ForwardingRuleResource: forwardingRule,
+			Region:                 a.Location,
+			Project:                a.projectID,
+		}
+		op, err = a.forwardingRulesClient.Insert(ctx, req)
+	}
+	if err != nil {
+		return fmt.Errorf("creating ComputeForwardingRule %s failed: %w", a.fullyQualifiedName(), err)
+	}
+	log.V(2).Info("successfully created ComputeForwardingRule", "name", a.fullyQualifiedName())
+	// Get the created resource
+	created := &computepb.ForwardingRule{}
```

Comment:
The `Insert` returns a `computepb.Operation`, which `op.Status` could be either OpDone or OpPending [Code](https://github.com/googleapis/google-cloud-go/blob/main/container/container.go#L45-L56). I think we shall probably only proceed the following GCP `Get` when we get a OpDone. 

For OpPending, we probably want to return  quietly and update the KCC `.status` in the next Reconcile, maybe `Find/Update`? @justinsb I'll vote for Update, do you have a preference?

Reply:
I won't worry about this because we call `op.[Wait](https://source.corp.google.com/piper///depot/google3/third_party/golang/cloud_google_com/go/compute/v/v1/apiv1/operations.go;l=48?q=apiv1%2Foperations.go&sq=package:piper%20file:%2F%2Fdepot%2Fgoogle3%20-file:google3%2Fexperimental)()` a few lines above, and it makes sure that the operation is Done.